internal/search/ag.go

```diff
@@ -0,0 +1,130 @@
+package search
+
+import (
+	"errors"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"github.com/launchdarkly/ld-find-code-refs/internal/log"
+	"github.com/launchdarkly/ld-find-code-refs/internal/validation"
+)
+
+type AgClient struct {
+	workspace string
+}
+
+func NewAgClient(path string) (*AgClient, error) {
+	if !filepath.IsAbs(path) {
+		log.Error.Fatalf("expected an absolute path but received a relative path: %s", path)
+	}
+	_, err := exec.LookPath("ag")
+	if err != nil {
+		return nil, errors.New("ag (The Silver Searcher) is a required dependency, but was not found in the system PATH")
+	}
+
+	return &AgClient{workspace: path}, nil
+}
+
+/*
+agSearchRegex splits search result lines into groups
+Group 1: File path.
+Group 2: Separator. A colon indicates a match, a hyphen indicates a context lines
+Group 3: Line number
+Group 4: Line contents
+*/
+var agSearchRegex = regexp.MustCompile("([^:]+)(:|-)([0-9]+)[:-](.*)")
+
+func (c *AgClient) searchForRefs(searchTerms []string, aliases map[string][]string, ctxLines int, delimiters []byte) (SearchResultLines, error) {
+	args := []string{"--nogroup", "--case-sensitive"}
+	pathToIgnore := filepath.Join(c.workspace, ignoreFileName)
+	if validation.FileExists(pathToIgnore) {
+		log.Debug.Printf("excluding files matched in %s", ignoreFileName)
+		args = append(args, fmt.Sprintf("--path-to-ignore=%s", pathToIgnore))
+	}
+	if ctxLines > 0 {
+		args = append(args, fmt.Sprintf("-C%d", ctxLines))
+	}
+
+	searchPattern := generateSearchPattern(searchTerms, delimiters, runtime.GOOS == windows)
+	/* #nosec */
+	cmd := exec.Command("ag", args...)
+	cmd.Args = append(cmd.Args, searchPattern, c.workspace)
+	out, err := cmd.CombinedOutput()
+	res := string(out)
+	if err != nil {
+		if err.Error() == "exit status 1" {
+			return nil, nil
+		} else if strings.Contains(res, searchTooLargeErr.Error()) ||
+			(runtime.GOOS == windows && strings.Contains(err.Error(), windowsSearchTooLargeErr.Error())) {
+			return nil, searchTooLargeErr
+		}
+		return nil, errors.New(res)
+	}
+
+	searchRegexWithFilteredPath, err := regexp.Compile("(?:" + regexp.QuoteMeta(c.workspace) + "/)" + agSearchRegex.String())
+	if err != nil {
+		return nil, err
+	}
+
+	output := string(out)
+	if runtime.GOOS == windows {
+		output = fromWindows1252(output)
+	}
+
+	ret := searchRegexWithFilteredPath.FindAllStringSubmatch(output, -1)
+
+	return c.generateReferences(aliases, ret, ctxLines, string(delimiters)), err
```

Comment:
Minor change that shouldn't change output - this method is now called to convert raw output lines to `SearchResultLines` for each (potentially paginated) search output. This is done to have the `SearchClient` interface conform to a standard interface returning `SearchResultLines`